api/odl.py

```diff
@@ -890,3 +956,509 @@ def _url_for(self, *args, **kwargs):
+            raise CannotHold()
+        start = datetime.datetime.strptime(availability.get("since"), self.TIME_FORMAT)
+        end = datetime.datetime.strptime(availability.get("until"), self.TIME_FORMAT)
+        position = entry.get("opds_holds", {}).get("position")
+        if position:
+            position = int(position)
+        # Get the loan base url from a link.
+        info_links = [link for link in entry.get("links") if link.get("rel") == "self"]
+        external_identifier = info_links[0].get("href")
+
+        return HoldInfo(
+            licensepool.collection,
+            licensepool.data_source.name,
+            licensepool.identifier.type,
+            licensepool.identifier.identifier,
+            start,
+            end,
+            hold_position=position,
+            external_identifier=external_identifier,
+        )
+
+    def release_hold(self, patron, pin, licensepool):
+        _db = Session.object_session(patron)
+
+        hold = get_one(
+            _db, Hold,
+            license_pool_id=licensepool.id,
+            patron=patron,
+        )
+        if not hold:
+            raise NotOnHold()
+
+        info_url = hold.external_identifier
+        try:
+            response = self._get(info_url)
+        except RemoteIntegrationException, e:
+            raise CannotReleaseHold()
+        feed = feedparser.parse(unicode(response.content))
+        entries = feed.get("entries")
+        if len(entries) < 1:
+            raise CannotReleaseHold()
+        entry = entries[0]
+        availability = entry.get("opds_availability", {})
+        if availability.get("status") not in ["reserved", "ready"]:
+            raise CannotReleaseHold()
+        revoke_links = [link for link in entry.get("links") if link.get("rel") == "http://librarysimplified.org/terms/rel/revoke"]
+        if len(revoke_links) < 1:
+            raise CannotReleaseHold()
+        revoke_url = revoke_links[0].get("href")
+        try:
+            self._get(revoke_url)
+        except RemoteIntegrationException, e:
+            raise CannotReleaseHold()
+        return True
+
+    def patron_activity(self, patron, pin):
+        _db = Session.object_session(patron)
+        loans = _db.query(Loan).join(Loan.license_pool).filter(
+            LicensePool.collection_id==self.collection_id
+        ).filter(
+            Loan.patron==patron
+        )
+
+        holds = _db.query(Hold).join(Hold.license_pool).filter(
+            LicensePool.collection_id==self.collection_id
+        ).filter(
+            Hold.patron==patron
+        )
+        
+        activity = []
+        for loan in loans:
+            info_url = loan.external_identifier
+            response = self._get(info_url, allowed_response_codes=["2xx", "3xx", "400"])
+            if response.status_code == 400:
+                # 400 is returned when the loan has been deleted. Leave this loan out of the result.
```

Comment:
400 seems weird--I would expect a 404 or 410. What's going on here?

Reply:
It's the NO_ACTIVE_LOAN problem detail.

Reply:
Maybe it doesn't make sense to reuse that when we have a link to a specific loan.

Reply:
Yeah, NO_ACTIVE_LOAN is more like "you tried to do something to a book that required an active loan as a prerequisite, but you don't have an active loan", this is more like "you asked for information about a specific loan that doesn't exist".